### PR TITLE
Provide First-Come-First-Remembered security model

### PR DESCRIPTION
**Pull request by chrysn**
Closes: https://github.com/core-wg/resource-directory/issues/254

### CODE REVIEW

**Comment by chrysn** [comment]
I'd appreciate a pair of eyes on this before tomorrow's interim, given my [original scribbling](https://github.com/core-wg/resource-directory/issues/254#issuecomment-690286601) was flawed in several respects, and I'm not sure I got all the fixes right.

**Comment by chrysn** [comment]
A question to be raised today is whether this should contain a note that where one wants endpoint names from certificates ("the other proposal" from #254), they can write it up per application or make a short document of it.

File: draft-ietf-core-resource-directory.md [hunk 1]
```diff
@@ -1703,6 +1703,51 @@ Care must be taken in such setups to determine the applicable access control mea
 One easy way to do that is to mandate the use of the sector parameter on all operations,
 as no credentials are suitable for operations across sector borders anyway.
 
+## First-Come-First-Remembered: A default policy
+
+The Fist-Come-First-Remembered policy is provided both as a reference example for a security policy definition,
```
Comment (cabo) [inline draft-ietf-core-resource-directory.md:1708]:
```suggestion
The First-Come-First-Remembered policy is provided both as a reference example for a security policy definition,
```

File: draft-ietf-core-resource-directory.md [hunk 2]
```diff
@@ -1703,6 +1703,51 @@ Care must be taken in such setups to determine the applicable access control mea
...
+
+Under this policy, the RD accepts registrations for any endpoint name that is not assigned to an active registration resource,
+and only accepts registration updates from the same endpoint.
+The policy is minimal in that towards lookup clients it does not make any of the claims of {{entered-resources}} and {{link-confidentiality}},
+and its claims on {{secure-ep}} are limited to the lifetime of that endpoint's registration.
+It does, however, guarantee towards any endpoint that for the duration of its registration, its links will be discoverable on the RD.
+
+When a registration or operation is attempted, the RD MUST determine the client's subject name or public key:
+
+* If the client's credentials indicate a subject name certified by an authority the RD recognizes (which may be the system's trust anchor store), that subject name is stored.
```
Comment (cabo) [inline draft-ietf-core-resource-directory.md:1720]: which == authority?
Maybe rephrase to make this less ambiguous.

File: draft-ietf-core-resource-directory.md [hunk 3]
```diff
@@ -1703,6 +1703,51 @@ Care must be taken in such setups to determine the applicable access control mea
...
+  With COSE based credentials (like in ACE), the name is taken from the Subject (sub) claim; this step fails if no Subject claim is part of the token.
```
Comment (cabo) [inline draft-ietf-core-resource-directory.md:1721]: COSE ≠ CWT; is this about CWT?
"the token" is what (the CWT)?

File: draft-ietf-core-resource-directory.md [hunk 4]
```diff
@@ -1703,6 +1703,51 @@ Care must be taken in such setups to determine the applicable access control mea
...
+  With X.509 certificates, the Common Name (CN) and the complete list of SubjectAltName entries are stored.
+  In both cases, the authority that certified the claim is stored along the subject, as the latter may only be locally unique.
```
Comment (cabo) [inline draft-ietf-core-resource-directory.md:1723]:
```suggestion
  In both cases, the authority that certified the claim is stored along with the subject, as the latter may only be locally unique.
```

File: draft-ietf-core-resource-directory.md [hunk 5]
```diff
@@ -1703,6 +1703,51 @@ Care must be taken in such setups to determine the applicable access control mea
...
+* Otherwise, if the client's credentials contain a public key, that public key is stored.
```
Comment (cabo) [inline draft-ietf-core-resource-directory.md:1724]: I think we are assuming that the client is proving possession of the private key for that public key.

File: draft-ietf-core-resource-directory.md [hunk 6]
```diff
@@ -1703,6 +1703,51 @@ Care must be taken in such setups to determine the applicable access control mea
...
+  With (D)TLS, this applies both to raw public keys and to the public keys indicated in certificates that are not recognized by the RD.
```
Comment (cabo) [inline draft-ietf-core-resource-directory.md:1725]: certificates that are not recognized?  How do these work?

File: draft-ietf-core-resource-directory.md [hunk 7]
```diff
@@ -1703,6 +1703,51 @@ Care must be taken in such setups to determine the applicable access control mea
...
+* If neither is present, a reference to the security session itself is stored.
+  With (D)TLS, that is the connection itself, or the session resumption information if available.
+  With OSCORE, that is the security context.
+
+As part of the registration operation, that information is stored along the registration resource.
+
+The RD MUST accept all registrations that do not conflict with an already active registration resource,
```
Comment (cabo) [inline draft-ietf-core-resource-directory.md:1732]: define conflict

File: draft-ietf-core-resource-directory.md [hunk 8]
```diff
@@ -1703,6 +1703,51 @@ Care must be taken in such setups to determine the applicable access control mea
...
+as long as they are made using a security layer supported by the RD.
+
+Any operation on a registration resource,
+including registrations that lead to an existing registration resource,
+MUST be rejected by the RD unless the credentials information is identical to (or, in case of SubjectAltNames, a superset of) the stored information.
```
Comment (cabo) [inline draft-ietf-core-resource-directory.md:1737]: The whole credentials information or the stored name?
(I'm not sure that "the" credentials information is defined here, either.)

File: draft-ietf-core-resource-directory.md [hunk 9]
```diff
@@ -1703,6 +1703,51 @@ Care must be taken in such setups to determine the applicable access control mea
...
+
+Note that even though subject names are compared in this policy,
+they are never directly compared to endpoint names,
+and an endpoint can not expect to "own" any particular endpoint name outside of an acive registration --
```
Comment (cabo) [inline draft-ietf-core-resource-directory.md:1741]:
```suggestion
and an endpoint can not expect to "own" any particular endpoint name outside of an active registration --
```

File: draft-ietf-core-resource-directory.md (hunk 2 above)
Comment (chrysn) [reply draft-ietf-core-resource-directory.md]: How about "indicate a subject name that is certified by any authority which the RD recognizes" as in the latest changeset? (I'm trying to avoid saying "contains a certificate from for" here because the certificate may be pointed at, and to leave the form of the certified subject name open for the next paragraphs to provide details.)

File: draft-ietf-core-resource-directory.md (hunk 3 above)
Comment (chrysn) [reply draft-ietf-core-resource-directory.md]: Yes, CWT (or JWT) is what I meant. This also means it's not "like in ACE" but "as common with ACE". Then, "the token" also makes sense.

File: draft-ietf-core-resource-directory.md (hunk 5 above)
Comment (chrysn) [reply draft-ietf-core-resource-directory.md]: Yes; rephrased.

File: draft-ietf-core-resource-directory.md (hunk 6 above)
Comment (chrysn) [reply draft-ietf-core-resource-directory.md]: I didn't mean unknown types of certificates but more certificates not recognized by any of the abovementioned authorities; rephrased.

File: draft-ietf-core-resource-directory.md (hunk 8 above)
Comment (chrysn) [reply draft-ietf-core-resource-directory.md]: Mh, I had the extraction steps phrased as "MUST determine ...", and considered changing that to "MUST extract" so I can refer to "the extracted information" here -- but that's not quite it: If the client originally just had a certificate without a good trust chain (so the public key was stored), and later comes with a good certificate, the key should still be enough, so it became:

"unless all the stored information is found in the new request's credentials."

**Comment by chrysn** [comment]
Looking into the COSE!=CWT more deeply, I now think that they are not even mutually exclusive (TLS X.509 client certificate with SubjectAltName and a CWT posted to /token with a subj claim? TLS with CWT as certificate, and an X.509 posted to token? Reminds me of IP-over-DNS-over-HTTP-over-QUIC. Not that I'd do that...), so I'll send this through one more iteration that just takes the bag of all found (subject, authority) claims and requires them. Might make phrasing it easier, even.
